File: app_main/src/test/java/com/github/fj/restapi/test/HttpRequestUtils.java
/*
 * springboot-multimodule-kotlin skeleton.
 * Under no licences and warranty.
 */
package com.github.fj.restapi.test;

import javax.servlet.http.HttpServletRequest;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

/**
 * @author Francesco Jo(nimbusob@gmail.com)
 * @since 31 - Oct - 2018
 */
public final class HttpRequestUtils {
    public static HttpServletRequest newMockHttpServletRequestByLocalhost() {
        final HttpServletRequest mocked = mock(HttpServletRequest.class);
        when(mocked.getRemoteAddr()).thenReturn("localhost");

        return mocked;
    }
}

File: _application/src/main/kotlin/com/github/fj/restapi/util/HttpRequestUtils.kt
/*
 * springboot-multimodule-kotlin skeleton.
 * Under no licences and warranty.
 */
package com.github.fj.restapi.util

import com.github.fj.lib.text.isNullOrUnicodeBlank
import javax.servlet.http.HttpServletRequest

/**
 * This logic searches for "X-Real-IP" header first and will use the value if query was hit,
 * otherwise try on "X-forwarded-for" header as callback and use the value if found,
 * and will use [HttpServletRequest.getRemoteAddr] for last resort which is not accurate under
 * reverse proxy configuration.
 *
 * Note that this approach is not accurate on various web containers and additional logic may be
 * needed to fulfill your business requirement.
 */
fun HttpServletRequest.extractIp(): String {
    with (getHeader("X-Real-IP")) {
        if (!isNullOrUnicodeBlank() && !isUnknown()) {
            return this
        }
    }

    with (getHeader("Proxy-Client-IP")) {
        if (!isNullOrUnicodeBlank() && !isUnknown()) {
            return this
        }
    }

    with (getHeader("X-forwarded-for")) {
        if (!isNullOrUnicodeBlank()) {
            split(",").let {
                if (!it[0].isUnknown()) {
                    return it[0]
                }
            }
        }
    }

    return remoteAddr
}

private fun String.isUnknown() = "unknown".equals(this, true)